File: java-source/src/main/java/com/example/api/IOUApi.java
package com.example.api;

import com.example.contract.IOUContract;
import com.example.flow.IOUFlow;
import com.example.state.IOUState;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.crypto.Party;
import net.corda.core.messaging.CordaRPCOps;
import net.corda.core.transactions.SignedTransaction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.concurrent.ExecutionException;

// This API is accessible from /api/example. All paths specified below are relative to it.
@Path("example")
public class IOUApi {
    private final CordaRPCOps services;

    static private final Logger logger = LoggerFactory.getLogger(IOUApi.class);

    public IOUApi(CordaRPCOps services) {
        this.services = services;
    }

    /**
     * Displays all IOU states that exist in the node's vault.
     */
    @GET
    @Path("ious")
    @Produces(MediaType.APPLICATION_JSON)
    public List<StateAndRef<ContractState>> getIOUs() { return services.vaultAndUpdates().getFirst(); }

    /**
     * Initiates a flow to agree an IOU between two parties.
     *
     * Once the flow finishes it will have written the IOU to ledger. Both the sender and the recipient will be able to
     * see it when calling /api/example/ious on their respective nodes.
     *
     * This end-point takes a Party name parameter as part of the path. If the serving node can't find the other party
     * in its network map cache, it will return an HTTP bad request.
     *
     * The flow is invoked asynchronously. It returns a future when the flow's call() method returns.
     */
    @PUT
    @Path("{party}/create-iou")
    public Response createIOU(Integer iouValue, @PathParam("party") String partyName) throws InterruptedException, ExecutionException {
        final Party otherParty = services.partyFromName(partyName);

        if (otherParty == null) {
            return Response.status(Response.Status.BAD_REQUEST).build();
        }

        final IOUState state = new IOUState(
                iouValue,
                services.nodeIdentity().getLegalIdentity(),
                otherParty,
                new IOUContract());

        Response.Status status;
        String msg;
        try {
            // The line below blocks and waits for the flow to return.
            final SignedTransaction result = services
                    .startFlowDynamic(IOUFlow.Initiator.class, state, otherParty)
                    .getReturnValue()
                    .get();

            status = Response.Status.CREATED;
            msg = String.format("Transaction id %s committed to ledger.", result.getId());

        } catch (Throwable ex) {
            status = Response.Status.BAD_REQUEST;
            msg = "Transaction failed.";
            logger.error(ex.getMessage(), ex);
        }

        return Response
                .status(status)
                .entity(msg)
                .build();
    }
}

File: kotlin-source/src/main/kotlin/com/iou/kotlin/api/IOUApi.kt
package com.iou.kotlin.api

import com.iou.kotlin.contract.IOUContract
import com.iou.kotlin.flow.IOUFlow
import com.iou.kotlin.flow.IOUSettleFlow
import com.iou.kotlin.flow.SelfIssueCashFlow
import com.iou.kotlin.state.IOUState
import net.corda.contracts.asset.Cash
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.messaging.CordaRPCOps
import net.corda.core.messaging.startFlow
import java.util.concurrent.ExecutionException
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

// This API is accessible from /api/iou. All paths you specify are relative to this root.
@Path("iou")
class IOUApi(private val services: CordaRPCOps) {
    /**
     * Displays all IOU states that exist in the node's vault.
     */
    @GET
    @Path("ious")
    @Produces(MediaType.APPLICATION_JSON)
    fun getIOUs() = services.vaultAndUpdates().first.filter { it.state.data is IOUState }

    /**
     * Displays all cash states that exist in the node's vault.
     */
    @GET
    @Path("cash")
    @Produces(MediaType.APPLICATION_JSON)
    fun getCash() = services.vaultAndUpdates().first.filter { it.state.data is Cash.State }

    /**
     * Initiates a flow to agree an IOU between two parties.
     */
    @GET
    @Path("create-iou")
    fun createIOU(
            @QueryParam(value = "value") iouValue: Int,
            @QueryParam(value = "party") partyName: String): Response {

        val thisParty = services.nodeIdentity().legalIdentity
        val otherParty = services.partyFromName(partyName)!!

        val state = IOUState(iouValue, thisParty, otherParty, IOUContract())

        // The line below blocks and waits for the flow to return.
        val result = services
                .startFlowDynamic(IOUFlow.Initiator::class.java, state, otherParty)
                .returnValue
                .get()

        return Response
                .status(Response.Status.CREATED)
                .entity("Transaction id ${result.id} committed to ledger.").build()
    }

    /**
     * Helper end-point to issue some cash to ourselves.
     */
    @GET
    @Path("self-issue-cash")
    @Produces(MediaType.APPLICATION_JSON)
    fun selfIssueCash(@QueryParam(value = "amount") amount: Int): Response {
        val cashState = services.startFlowDynamic(SelfIssueCashFlow::class.java, amount).returnValue.get()
        return Response.status(Response.Status.CREATED).entity(cashState.toString()).build()
    }

    /**
     * Settles an IOU. Requires cash to be able to settle.
     */
    @GET
    @Path("settle-iou")
    @Produces(MediaType.APPLICATION_JSON)
    fun settleIOU(@QueryParam(value = "id") id: String,
                  @QueryParam(value = "amount") amount: Int): Response {
        val linearId = UniqueIdentifier.fromString(id)
        services.startFlowDynamic(IOUSettleFlow.Initiator::class.java, linearId, amount).returnValue.get()
        return Response.status(Response.Status.CREATED).entity("$amount USD paid off on IOU id $id.").build()
    }
}